packages/vue-app/template/mixins/fetch.server.js

```diff
@@ -0,0 +1,44 @@
+import Vue from 'vue'
+import { hasFetch } from '../utils'
+
+function getDataDiff(o1, o2) {
+  return Object.keys(o2).reduce((diff, key) => {
+    if (o1[key] === o2[key]) {
+      return diff
+    }
+    return {
+      ...diff,
+      [key]: o2[key]
+    }
+  }, {})
+}
+
+export default {
+  beforeCreate () {
+    if (hasFetch(this)) {
+      this._fetchOnServer = this.$options.fetchOnServer !== false
+      this.$isFetching = !this._fetchOnServer
+      Vue.util.defineReactive(this, '$fetchError', null)
+    }
+  },
+  async serverPrefetch () {
+    if (hasFetch(this) && this._fetchOnServer) {
+      const data = Object.assign({}, this.$data)
+
+      try {
+        await this.$options.fetch.call(this)
+      } catch (err) {
+        this.$fetchError = err
+      }
+      // Define and ssrKey for hydration
+      this._ssrKey = this.$ssrContext.nuxt.data.length
```

Comment:
@Atinux can you give a little more detail on why this works? At first glance it seems prone to collisions?

Reply:
Well it's to generate the `data-ssr-key` to keep `index` of `window.__NUXT__.data` when hydrating, since the **same component** with `fetch` can be used multiple times.

Reply:
But why is it data.length and not some hash of the data?

Reply:
`nuxt.data` is actually an array, not sure to understand what you mean here

Reply:
@bf component `asyncData` objects are pushed to `window.__NUXT__` in the order they're loaded, so that's why using `data.length` as a unique per-component key works.